File: test/native/AlarmLaunchPayloadTest.swift
import Foundation

private final class FailedSyncDefaults: UserDefaults {
  override func synchronize() -> Bool { false }
}

private final class FailedReadbackDefaults: UserDefaults {
  override func synchronize() -> Bool { true }
  override func dictionary(forKey defaultName: String) -> [String: Any]? {
    ["scheduleId": "fixture-schedule", "fingerprint": "private-marker"]
  }
}

@main
struct AlarmLaunchPayloadTest {
  static func main() {
    let marker = "private-step-title-do-not-forward"
    let old: [String: Any] = [
      "scheduleId": "fixture-schedule", "fingerprint": marker,
      "title": marker, "body": marker, "alarmTime": "1234",
      "alarmLaunchAction": "startPreparation", "unknown": marker,
      "alarmLaunchPayloadVersion": "8", "type": "start-command"
    ]
    let expected = ["scheduleId": "fixture-schedule", "type": "schedule_alarm",
                    "alarmLaunchPayloadVersion": "9", "promptVariant": "alarm"]
    precondition(AlarmLaunchPayload.sanitize(old) == expected)
    precondition(AlarmLaunchPayload.sanitize(expected) == expected)
    for id: Any in [123, ["id"], "", "  ", "x\ny", "x\u{7f}y", String(repeating: "x", count: 513)] {
      precondition(AlarmLaunchPayload.sanitize(["scheduleId": id]) == nil)
    }
    precondition(AlarmLaunchPayload.sanitize(["scheduleId": "한글-🙂"]) != nil)
    let suite = "ontime-a11-test-\(UUID().uuidString)"
    let defaults = UserDefaults(suiteName: suite)!
    defer { defaults.removePersistentDomain(forName: suite) }
    let key = "on_time_alarm_launch_payload"
    defaults.set("durable-sentinel", forKey: "unrelated")
    defaults.set(old, forKey: key)
    precondition(AlarmLaunchPayload.sanitizeStored(in: defaults, key: key))
    precondition(defaults.dictionary(forKey: key) as? [String: String] == expected)
    precondition(!String(describing: defaults.object(forKey: key)).contains(marker))
    precondition(AlarmLaunchPayload.sanitizeStored(in: defaults, key: key))
    for bad: Any in [marker, ["fingerprint": marker], ["scheduleId": 123, "fingerprint": marker]] {
      defaults.set(bad, forKey: key)
      precondition(AlarmLaunchPayload.sanitizeStored(in: defaults, key: key))
      precondition(defaults.object(forKey: key) == nil)
      precondition(defaults.string(forKey: "unrelated") == "durable-sentinel")
    }
    precondition(AlarmLaunchPayload.sanitizeStored(in: defaults, key: key))
    let failedSync = FailedSyncDefaults(suiteName: suite)!
    precondition(!AlarmLaunchPayload.sanitizeStored(in: failedSync, key: key))
    let failedReadback = FailedReadbackDefaults(suiteName: suite)!
    precondition(!AlarmLaunchPayload.sanitizeStored(in: failedReadback, key: key))
    print("iOS launch privacy contract passed (legacy, malformed, idempotent, stored cleanup)")
  }
}

File: test/native/AlarmLaunchPayloadTest.kt
package club.devkor.ontime

fun main() {
    val privateMarker = "private-step-title-do-not-forward"
    val old = mapOf(
        "scheduleId" to "fixture-schedule", "fingerprint" to privateMarker,
        "title" to privateMarker, "body" to privateMarker,
        "alarmTime" to "1234", "preparationStartTime" to "1200",
        "nativeAlarmId" to "42", "alarmLaunchAction" to "startPreparation",
        "alarmLaunchPayloadVersion" to "8", "unknown" to privateMarker,
        "type" to "start-command", "promptVariant" to "bypass-confirmation",
    )
    val clean = requireNotNull(AlarmLaunchPayload.sanitize(old))
    check(clean == mapOf(
        "scheduleId" to "fixture-schedule", "type" to "schedule_alarm",
        "alarmLaunchPayloadVersion" to "10", "promptVariant" to "alarm",
    ))
    check(!clean.toString().contains(privateMarker))
    check(AlarmLaunchPayload.sanitize(clean) == clean)
    for (id in listOf(null, 5, listOf("id"), "", "  ", "x\ny", "x\u007fy", "x".repeat(513))) {
        check(AlarmLaunchPayload.sanitize(mapOf("scheduleId" to id)) == null)
    }
    check(AlarmLaunchPayload.sanitize(mapOf("scheduleId" to "한글-🙂")) != null)
    val identity = "11111111-1111-1111-1111-111111111111"
    val issued = old + mapOf("storeIncarnation" to identity)
    check(AlarmLaunchPayload.sanitize(issued)?.get("storeIncarnation") == identity)
    check(AlarmLaunchPayload.sanitize(AlarmLaunchPayload.deliveryExtras(issued))?.get("storeIncarnation") == identity)
    check(!clean.containsKey("storeIncarnation"))
    for (bad in listOf(42, "", "private-value", "g".repeat(36), null)) {
        check(AlarmLaunchPayload.sanitize(old + mapOf("storeIncarnation" to bad)) == null)
    }
    val delivery = AlarmLaunchPayload.deliveryExtras(old)
    check(delivery["title"] == privateMarker && delivery["body"] == privateMarker)
    check(delivery["nativeAlarmId"] == "42" && delivery["alarmTime"] == "1234")
    check(!delivery.containsKey("fingerprint") && !delivery.containsKey("alarmLaunchAction"))
    check(AlarmLaunchPayload.sanitize(delivery) == clean)
    val malformed = AlarmLaunchPayload.deliveryExtras(mapOf(
        "scheduleId" to 123, "title" to privateMarker, "nativeAlarmId" to 42,
        "alarmTime" to "not-a-time", "preparationStartTime" to 2.5,
    ))
    // Cancellation ownership survives malformed routing; content does not.
    check(malformed == mapOf("nativeAlarmId" to "42"))
    println("Android launch privacy contract passed (legacy, malformed, idempotent, delivery separation)")
}
